feat(quantic): Case Classification Component

mmitiche opened the pull request:
# Case Classification Component

lbergeron commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,273 @@
+   * Returns the options, excluding the suggestions.
+   * @returns {Array<object>}
+   *  */
+  get filtredOptions() {
+    return this.options.filter((option) => {
+      return (
+        this.classifications.findIndex(
+          (item) => item.value === option.value
+        ) === -1
+      );
+    });
+  }
```
```suggestion
  get filteredOptions() {
    return this.options.filter((option) => 
      !this.classifications.some((item) => item.value === option.value)
    );
  }
```

lbergeron reviewed (commented):
I didn't review all the changes, mostly did a quick check on the component itself. The Cypress tests are not running on my side though. I guess it is because this task is not yet completed (https://coveord.atlassian.net/browse/SFINT-4327).

mmitiche commented:
> I didn't review all the changes, mostly did a quick check on the component itself. The Cypress tests are not running on my side though. I guess it is because this task is not yet completed (https://coveord.atlassian.net/browse/SFINT-4327).

Exactly that's why, I'm currently working on that Jira, so the tests will be good to run soon 👌

lbergeron commented:
> > I didn't review all the changes, mostly did a quick check on the component itself. The Cypress tests are not running on my side though. I guess it is because this task is not yet completed (https://coveord.atlassian.net/browse/SFINT-4327).
> 
> Exactly that's why, I'm currently working on that Jira, so the tests will be good to run soon ok_hand

Perfect, thanks! I updated the ticket accordingly.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,273 @@
+   * @returns {boolean}
+   */
+  get isMoreOptionsVisible() {
+    return this.options.length > Math.max(Number(this.maxChoices), 1);
```
Actually, I don't want maxChoices to be equal to 0.  by a choice I mean any kind of option that we can propose to the user, if for example we display only the select dropdown  that's counted as 1 choice. if we display one suggestions and the select dropdown that's counted as 2 choices and so on . so for this reason MaxChoices can be only equal or greater than 1 cause we always at least want to show a choice to the user.

lbergeron replied:
Ah OK I see. My understanding was that `maxChoices` was the maximum number of values displayed next to the dropdown.

erocheleau replied:
I have the same understanding as Luc for this option, so perhaps a better name is required?

lbergeron replied:
`maxSuggestions` maybe? Setting it to zero would mean "no suggestions", and thus only the drop-down would be displayed.

nathanlb reviewed (commented):
I haven't reviewed the test suite yet but did a pass on the rest. This is nice work!

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,285 @@
+  @api engineId;
+
+  /**
+   * The field of the case to be classified.
```
Nope, I just forgot to add them.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,285 @@
+   * Hide the suggested options.
+   * @returns {void}
+   */
+  _hideSuggestions() {
```
I prefixed it that way only cause its a private method used only internally by this component. I think I will just remove it since we are not following such conventions.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,285 @@
+    const suggestions = this.template.querySelectorAll('.slds-visual-picker');
+    suggestions.forEach((suggestion) => {
+      // @ts-ignore
+      suggestion.style.width = `${suggestion.clientWidth}px`;
```
Yes It would be simpler to perform the hiding using the html conditional templates  and without the animation effect. The UX team preferred to have this animation  effect so that's why it's here,  is there any reason we want to avoid having animations effects? or is it a question of having this component consistant with the other quantic components, like I understand that having all the other Quantic components without any animation effects and this one containing one would be a bit weird.

mmitiche commented on packages/quantic/cypress/page-objects/search.ts:
```diff
@@ -131,6 +139,12 @@ export function interceptSearchIndefinitely(): {sendResponse: () => void} {
   return interceptIndefinitely(routeMatchers.search);
 }
 
+export function interceptClassificationsIndefinitely(): {
```
Absolutely👌 I will separate all the new additions in a new file like you suggested.

erocheleau commented on packages/quantic/cypress/integration/case-classification/case-classification.cypress.ts:
```diff
@@ -0,0 +1,385 @@
+  messageWhenValueMissing: string;
+}
+
+describe('quantic-case-classification', () => {
```
It might be a stupid test to do, but does it make sense to test that the suggestions appear in the order that they were received from the API?

I only see tests that checks the correct number of suggestions are displayed for example. Would it make sense to also test in the order in which they get displayed?

mmitiche replied:
Absolutely ! I will add such tests 👌

erocheleau commented on packages/quantic/force-app/examples/main/lwc/exampleQuanticCaseClassification/exampleQuanticCaseClassification.js:
```diff
@@ -0,0 +1,58 @@
+
+export default class ExampleQuanticCaseClassification extends LightningElement {
+  @api engineId = 'case-assist-engine';
+  @api caseAssistId = 'a4fb453a-b1f1-4054-9067-bef117586baa';
```
What do we think of adding perhaps a utils function to `getDefaultCaseAssistId()` instead of hardcoding a string value directly here?

erocheleau replied:
On second thought it might make sense to keep it there and contained.

mmitiche replied:
Is there an API call that I can make to retrieve the case assist config ID ?

nathanlb replied:
[Here ](https://platformdev.cloud.coveo.com/docs/?urls.primaryName=Customer%20Service#/Configurations/getAllCaseAssistConfigurations)are all the case assist API calls. Yes there is one that lists all configs in an org. Keep in mind that the token needs certain privileges to successfully get this info. (It might already)

mmitiche replied:
Great! Thank you💪

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.css:
```diff
@@ -0,0 +1,37 @@
+  --sds-c-icon-color-foreground-default: white;
+}
+.slds-visual-picker__figure{
+  width: 9.875rem;
```
Did you take these values somewhere from a component in the SLDS?

Seems weird to have such fractions of rem;

mmitiche replied:
Nope, these values came from the design mockups, I used the same width used in the design and  converting it  from px to rem gave such fractions.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.css:
```diff
@@ -0,0 +1,37 @@
+  height: 1.938rem;
+}
+
+.loading-holder{
```
maybe linter needs to run on this file to have the `{` with a space after the class?

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.html:
```diff
@@ -0,0 +1,65 @@
+          <template for:each={suggestions} for:item="option">
+            <div key={option.value}
+              class="case-classification-suggestion slds-visual-picker slds-visual-picker_medium slds-m-left_none slds-var-m-right_medium slds-var-m-top_x-small">
+              <input type="radio" data-suggestion-id={option.id} onchange={handleSelectClassification} id={option.value} value={option.value} name='options' />
```
```suggestion
              <input type="radio" data-suggestion-id={option.id} onchange={handleSelectClassification} id={option.value} value={option.value} name='suggestions' />
```

mmitiche replied:
The suggestions and the inline options are both displayed using radio inputs and that's why I used the name value `options`, to have all the radio inputs with the same name. otherwise if the suggestions are named `suggestions` and the inline options are named `options` the user will have the possibility to choose more than one option.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.html:
```diff
@@ -0,0 +1,65 @@
+          </div>
+        </template>
+        <template if:false={loading}>
+          <template for:each={suggestions} for:item="option">
```
```suggestion
          <template for:each={suggestions} for:item="suggestion">
```
Just to be consistent? Do you prefer that name?

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,299 @@
+  @wire(getObjectInfo, {objectApiName: CASE_OBJECT})
+  objectInfo;
+
+  @wire(getPicklistValuesByRecordType, {
```
This gets ALL values for ALL picklist fields?

Is there a way we only get the picklist values for only this specific field?

`fieldName`

erocheleau replied:
```js
@api(getPicklistValues, { recordTypeId: '$objectInfo.data.defaultRecordTypeId', fieldApiName: CASE_OBJECT[fieldName]})
    picklistValues;
```

I am really not sure if this works, but maybe it does? 😂

erocheleau replied:
It's always been a bit fuzzy the order of execution of things if you modify a variable used in an @wire if the wire refreshes itself correctly

mmitiche replied:
Yes exactly, this will get the values  for ALL picklist fields. There is a way to to get only the picklist values for only one specific field but following the doc to achieve this import the field from Salesforce is needed, an import like the following:
`import PRIORITY_FIELD from '@salesforce/schema/Case.Priority';`
and the getting the piclist values like the following:
```
@wire(getPicklistValues, { recordTypeId: RECORD_ID, fieldApiName: PRIORITY_FIELD })
    propertyOrFunction;
```

But since in our case the field can be anything I can't do such import since I don't know the field that the user will want to classify, so I found this workaround of getting all the pickilist values and then using only the values of the field specified by the user.

erocheleau replied:
`import PRIORITY_FIELD from '@salesforce/schema/Case.Priority';`
Inspect what the import gives you at runtime, I *think* it might just be the API name of the field.
It might just be that `PRIORITY_FIELD` is `Priority` as a string :P 

Long time since I checked.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,285 @@
+  @api selectPlaceholder = this.labels.moreTopics;
+
+  /**
+   * The maximum number of choices to be displayed, a choice can be a suggestion, an inline option or the select dropdown.
```
I would like to understand better this option too.
According to the tests, I think `1, 0 and -1` all have the same effect basically?

mmitiche replied:
Exactly, in other words the minimum value  that `maxChoices` can take is 1.  The logic behind this is that max choices represents the maximum number of choices we propose to the user, and a single choice can be a suggestion, an inline option or the select drop down. so for example if maxChoices is set to 3 we can show 2 suggestions and the select drop down. if maxChoices is set to 1 we only can show the select drop down. Having maxChoices set to 0 means that we don't want anything no suggestion and no inline option and no select drop down and this will make our component kind of useless. so maxChoices is always greater or equal to 1. and in the case where the user gives maxChoices a negative value or 0, we always reset the value to 1.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,299 @@
+  @api engineId;
+
+  /**
+   * The field of the case to be classified.
```
You will have to be precise here if the fieldName is the salesforce API name, a label, or the Coveo field name

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,299 @@
+    this.engine = engine;
+    this.field = CoveoHeadlessCaseAssist.buildCaseField(engine, {
+      options: {
+        field: `sf${this.fieldName.toLowerCase()}`,
```
ouf, this is a discussion entirely...

I think it is wrong to assume that the field from the CaseAssist will always have this format. And as mentioned above on the `@api fieldName` property. You are playing with 3 different potential values here.

SF field API name => example `My_Custom_Field__c`

SF field label         => example `My Custom Field`

Coveo field name => example `sfmycustomfield` but can actually be ANYTHING the user wants. This is just the default value


I think I would do this:

```js
@api sfFieldAPIName;
@api coveoFieldName = `sf${this.sfFieldAPIName.toLowerCase()}`;
```

I am not even sure if we don't add a `c` at the end of the `sfmycustomfield` when it is a custom field in Salesforce
`sfmycustomfieldc`.

We used to by default, now I am not sure if it was removed.

mmitiche replied:
sfFieldAPIName and coveoFieldName added 👌

erocheleau reviewed (changes requested):
We can discuss more the field comment in january

mmitiche commented on packages/quantic/cypress/integration/case-classification/case-classification.cypress.ts:
```diff
@@ -0,0 +1,385 @@
+  }
+
+  describe('with default options', () => {
+    it('should work as expected', () => {
```
Absolutely👌

mmitiche commented on packages/quantic/cypress/integration/case-classification/case-classification.cypress.ts:
```diff
@@ -0,0 +1,385 @@
+        Expect.displaySelectInput(true);
+      });
+
+      scope('when selecting an option from the select input', () => {
```
I just added it👌

mmitiche commented on packages/quantic/cypress/integration/case-classification/case-classification.cypress.ts:
```diff
@@ -0,0 +1,385 @@
+  });
+
+  describe('with invalid number of maxChoices', () => {
+    it('should display only the select dropdown when maxChoices is set to 0', () => {
```
Yes
> Exactly, in other words the minimum value that `maxChoices` can take is 1. The logic behind this is that max choices represents the maximum number of choices we propose to the user, and a single choice can be a suggestion, an inline option or the select drop down. so for example if maxChoices is set to 3 we can show 2 suggestions and the select drop down. if maxChoices is set to 1 we only can show the select drop down. Having maxChoices set to 0 means that we don't want anything no suggestion and no inline option and no select drop down and this will make our component kind of useless. so maxChoices is always greater or equal to 1. and in the case where the user gives maxChoices a negative value or 0, we always reset the value to 1.

lbergeron commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,301 @@
+   * @type {boolean}
+   */
+  get isSelectVisible() {
+    return this._isSelectVisible || 0 >= this.suggestions.length;
```
Is there a particular reason why you put the constant on the left side of the condition? It is generally easier to read when putting the variable expression on the left and the constant expression on the right.

```suggestion
    return this._isSelectVisible || this.suggestions.length < 0;
```

lbergeron commented on packages/quantic/quantic-examples-community/profiles/Quantic Examples Profile.profile:
```diff
@@ -1,8 +1,483 @@
+    </fieldPermissions>
+    <fieldPermissions>
+        <editable>true</editable>
+        <field>Account.AnnualRevenue</field>
```
You are granting way too much accesses through this profile. Please keep the permissions to the absolute minimum for this profile.

erocheleau replied:
I think it's either we include it entirely or not at all?

Most of these are set to `false`

aelfaleh commented on packages/quantic/cypress/integration/case-classification/case-classification-actions.ts:
```diff
@@ -0,0 +1,22 @@
+
+function caseClassificationActions(selector: CaseClassificationSelector) {
+  return {
+    clickSuggestion: (idx: number) => selector.suggestedOption(idx).click(),
```
if you want you can add `logAction` to make your report more verbal

nathanlb commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,303 @@
+  }
+
+  updateFieldState() {
+    if(this.maxChoices > 1){
```
space after the if

nathanlb commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,303 @@
+   * @api
+   * @type {string}
+   */
+    @api coveoFieldName;
```
rogue space here

nathanlb commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,303 @@
+/** @typedef {import("coveo").CaseField} CaseField */
+
+/**
+ * The `QuanticCaseClassification` component displays field value suggestions returned by Coveo Case Assist as well as a single-select dropdown containing other available values.
```
We should add another sentence saying these other values in the dropdown are taken from the org's CASE_OBJECT record type.

nathanlb reviewed (changes requested):
This was clearer a lot of work and it's looking real good! It works nicely and the test suite is awesome.
A few comments and some points i'll write here:
1. The mock-ups say the classification with the highest degree of confidence should be selected automatically. This isn't the case here. Was it discussed with UX?
2. The height of the buttons and the select input is different from the mock-ups as well. It's fine as long as it was discussed with UX and the mock-ups gets updated so that QA can validate efficiently.
3. This should be adjusted just a bit so the colors match and the overlay doesn't stick out.
<img width="800" alt="Screen Shot 2022-01-04 at 3 15 29 PM" src="https://user-images.githubusercontent.com/16785453/148119135-992b730f-f404-4916-bfea-f535541de2d9.png">

mmitiche commented:
> This was clearer a lot of work and it's looking real good! It works nicely and the test suite is awesome. A few comments and some points i'll write here:
> 
> 1. The mock-ups say the classification with the highest degree of confidence should be selected automatically. This isn't the case here. Was it discussed with UX?
> 2. The height of the buttons and the select input is different from the mock-ups as well. It's fine as long as it was discussed with UX and the mock-ups gets updated so that QA can validate efficiently.
> 3. This should be adjusted just a bit so the colors match and the overlay doesn't stick out.
> 
> <img alt="Screen Shot 2022-01-04 at 3 15 29 PM" width="800" src="https://user-images.githubusercontent.com/16785453/148119135-992b730f-f404-4916-bfea-f535541de2d9.png">

1. The feature of having the classification with highest degree of confidence is now added.
2.  Yes the hight of the buttons and the select input was discussed with the UX, I sync with them to update the mockups like you mentioned 👌.
3. The overlay is now adjusted, no more stick out and it's now with the correct color:

https://user-images.githubusercontent.com/86681870/148387852-d64b31a4-6d56-401c-a8e7-e4d9fab4e99a.mov

nathanlb reviewed (approved):
👏 👏 👏

lvu285 commented:
Can we have an option to set number of Suggestion display?
for example, suggestion returns 4 values, and I want to display only first 2, the rest will be on the dropdown?

lvu285 commented:
Like our discussion
- Probably we can consider about the `maxChoice` naming, a bit confusing
- According to first discussion during our meeting, we should support user `uncheck` the box and submit it --> option need to be supported from Quantic

*Also, consider to have default value for number of suggestion shows _ maybe 3 suggestions + 1 dropdown (like in the mockups)

lvu285 commented on packages/quantic/cypress/integration/case-classification/case-classification.cypress.ts:
```diff
@@ -0,0 +1,424 @@
+    });
+  });
+
+  describe('when invalid number of maxChoices is given', () => {
```
would love to have a test to cover the situation when user set `invalid Salesforce field`

mmitiche commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -749,4 +749,32 @@
         <protected>false</protected>
         <shortDescription>Timeframe facet end label</shortDescription>
     </labels>
+    <labels>
```
@acarpentier01  please tell me if everything looks fine to you in the following labels.

mmitiche replied:
Absolutely! Thank you for your feedback👌

mmitiche commented on packages/quantic/force-app/main/translations/fr.translation-meta.xml:
```diff
@@ -400,4 +400,20 @@
         <label>Fin :</label>
         <name>quantic_EndLabel</name>
     </customLabels>
+    <customLabels>
```
@acarpentier01  french version of the labels.

erocheleau commented on packages/quantic/force-app/examples/main/lwc/utils/utils.js:
```diff
@@ -0,0 +1,24 @@
+// @ts-ignore
+import getHeadlessConfiguration from '@salesforce/apex/HeadlessController.getHeadlessConfiguration';
```
Is it really the best file to have this in?
At least perhaps a caseAssistUtils.js instead of a generic utils.js?

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,350 @@
+        const value = this.classifications.length
+          ? this.classifications[0].value
+          : '';
+        this.setValue(value);
```
Perhaps a slightly more evocative name? 

`setFieldValue` ?

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,350 @@
+        (option) =>
+          !this.classifications.some((item) => item.value === option.value)
+      )
+      .map((option) => {
+        if (option.value === this._value) {
+          return {...option, checked: true};
+        }
+        return {...option, checked: false};
+      });
```
```suggestion
      .map((option) => {
        return {...option, checked: (option.value === this._value)};
      });
```

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticCaseClassification/quanticCaseClassification.js:
```diff
@@ -0,0 +1,350 @@
+   * @returns {Array<object>}
+   */
+  get suggestions() {
+    return this.classifications.map((suggestion) => {
+      if (suggestion.value === this._value) {
+        return {...suggestion, checked: true};
+      }
+      return {...suggestion, checked: false};
+    });
```
```suggestion
    return this.classifications.map((suggestion) => {
      return {...suggestion, checked: (suggestion.value === this._value)};
    });
```

erocheleau commented on packages/quantic/quantic-examples-community/objects/Account.object:
```diff
@@ -0,0 +1,395 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This probably should not be in this PR or tracked?

erocheleau commented on packages/quantic/quantic-examples-community/objects/Case.object:
```diff
@@ -0,0 +1,421 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I don't think we should have that either, it's for giving access to the example community right?

Did you try with a permission set instead?

erocheleau replied:
What is the line exactly that grants the permissions? I see the permission on the profile you have included already, don't think we need the object on top of the profile

mmitiche replied:
True! I just retried to setup the community after removing the objects and the community was created correctly with the setting we needed. I will just double check cause I remember trying this previously and it didn't give me the needed results without including the objects. when I confirm that it's working correctly I will just delete the files related to the objects and I'm gonna keep only the profile metadata👌

erocheleau reviewed (approved):
Just probably not necessary to include Account object or Case object in that?
